ui/plugins/auto_classification/controller.js

```diff
@@ -69,891 +40,1104 @@ treeherder.factory('thStringOverlap', function() {
-                                                  cf.bug ? cf.bug.summary : "",
-                                                  cf.bug ? cf.bug.resolution : "",
-                                                  getClassificationMatches(cf.id));
+            options.forEach((option, idx) => {
+                option.hidden = idx > (minOptions - 1) &&
+                    (option.score < lowerCutoff ||
+                     option.score < bestRatio * bestScore ||
+                     idx > (maxOptions - 1));
             });
         }
 
-        function bugSuggestionOptions(data, autoOptions) {
-            // Add bugs matched by orangefactor but not autoclassifier
-            var autoBugs = autoOptions.reduce(function(classifiedBugs, x) {
-                classifiedBugs.add(x.bugNumber);
-                return classifiedBugs;
-            }, new Set());
+        /**
+         * Return a function that takes a classified failure id and returns the
+         * matcher that provided the best match, and the score of that match.
+         */
+        function getClassifiedFailureMatcher() {
+            var matchesByCF = line.data.matches.reduce(
+                function(matchesByCF, match) {
+                    if (!matchesByCF.has(match.classified_failure)) {
+                        matchesByCF.set(match.classified_failure, []);
+                    }
+                    matchesByCF.get(match.classified_failure).push(match);
+                    return matchesByCF;
+                }, new Map());
 
-            var bugSuggestions = data.unstructured_bugs
-                    .filter(function(bug) {
-                        return !autoBugs.has(bug.id);
+            return function(cf_id) {
+                return matchesByCF.get(cf_id).map(
+                    function(match) {
+                        return {
+                            matcher: ctrl.errorMatchers.get(match.matcher),
+                            score: match.score
+                        };
                     });
+            };
+        }
 
-            var message = [data.test, data.subtest, data.message]
-                    .filter(function(x) {return x !== null;})
-                    .join(" ");
-
-            var scores = bugSuggestions
-                .reduce(function(scores, bug) {
-                    var score = thStringOverlap(message, bug.summary);
-                    // Artificially reduce the score of resolved bugs
-                    score *= bug.resolution ? 0.8 : 1;
-                    scores.set(bug.id, score);
-                    return scores;
-                }, new Map());
+        /**
+         * Get the initial default option
+         * @param {Object[]} options - List of line-specific options
+         * @param {Object[]} extraOptions - List of line-independent options
+         * @param {Object[]} prevLine - Line before the current one in the log
+         */
+        function getDefaultOption(options, extraOptions, prevLine) {
```

Comment:
This function would really benefit from more comments.  You have some comments here, which helps, but the logic is pretty complex, so it'd be MUCH easier to review and understand later if you put in a play-by-play to some degree.

Something along the lines of: "Now we have determined that X is true, so we need to find out if we should Y or Z."  I may be the only one, but comments like that help me with a decision function such as this.  :)